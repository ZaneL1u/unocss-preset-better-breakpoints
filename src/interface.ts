export enum BreakpointEnum {
  /** extra large */
  XL = 'xl',
  /** large */
  LG = 'lg',
  /** medium */
  MD = 'md',
  /** small */
  SM = 'sm',
  /** extra small */
  XS = 'xs',
}

export type BreakpointType = `${BreakpointEnum}`

export type CSSUnit = 'vh' | 'vw' | 'px' | 'rem' | 'em' | 'pt' | 'pc' | 'mm' | 'cm' | 'in' | 'ex' | 'ch' | 'vmin' | 'vmax' | '%'

/**
 * Combined size unit
 *
 * @example
 * - '100px'
 * - '100%'
 * - '100vh'
 * - '100vw'
 */
export type CombinedSizeUnit = `${number}${CSSUnit}`

export enum BreakpointMatchTypeEnum {
  /** all */
  ALL = 'all',
  /** print */
  PRINT = 'print',
  /** screen */
  SCREEN = 'screen',
}

/**
 * Breakpoint option
 * @example
 * - 520
 * - '520px'
 * - { max: 520 }
 * - { max: 520, min: 1024 }
 * - { 'min-width': '520px' }
 * - { 'width': '1024px', 'orientation': 'landscape' }
 */
export type Breakpoint =
  | {
    max?: number | CombinedSizeUnit
    min?: number | CombinedSizeUnit
    orientation?: 'landscape' | 'portrait'
  } | {
    'max-width'?: number | CombinedSizeUnit
    'min-width'?: number | CombinedSizeUnit
    'max-height'?: number | CombinedSizeUnit
    'min-height'?: number | CombinedSizeUnit
    'width'?: number | CombinedSizeUnit
    'height'?: number | CombinedSizeUnit
    'orientation'?: 'landscape' | 'portrait'
  }
  | CombinedSizeUnit
  | number

export type BreakpointObj = Exclude<Breakpoint, string | number>

/**
 * Breakpoint option
 */
export type BreakpointOption =
  | { [BreakpointMatchTypeEnum.ALL]: Breakpoint, [BreakpointMatchTypeEnum.PRINT]?: never, [BreakpointMatchTypeEnum.SCREEN]?: never }
  | { [BreakpointMatchTypeEnum.PRINT]?: Breakpoint, [BreakpointMatchTypeEnum.SCREEN]?: Breakpoint, [BreakpointMatchTypeEnum.ALL]?: never }
  | Breakpoint

/**
 * Better breakpoints options
 *
 */
export type BetterBreakpointsOptions = {
  [key in BreakpointEnum | string]: BreakpointOption;
}
