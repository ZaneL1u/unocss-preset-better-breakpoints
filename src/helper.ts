import type { Breakpoint, BreakpointOption } from './interface'
import { BreakpointMatchTypeEnum } from './interface'

export function hasAnySpecifiedKey<T extends object>(keys: string[], obj: T) {
  return keys.some(key => Object.hasOwnProperty.call(obj, key))
};

export function isMatchType(option: BreakpointOption): option is { [BreakpointMatchTypeEnum.ALL]: Breakpoint } | { [BreakpointMatchTypeEnum.PRINT]: Breakpoint, [BreakpointMatchTypeEnum.SCREEN]: Breakpoint } {
  return typeof option === 'object' && hasAnySpecifiedKey(Object.values(BreakpointMatchTypeEnum), option)
}

export function convertKey(key: string): string {
  switch (key) {
    case 'min':
      return 'min-width'
    case 'max':
      return 'max-width'
    default:
      return key
  }
}
