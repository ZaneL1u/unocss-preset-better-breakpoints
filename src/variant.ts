import type { VariantHandler, VariantObject } from '@unocss/core'
import type { Theme } from '@unocss/preset-mini'
import type { BetterBreakpointsOptions, Breakpoint, BreakpointObj } from './interface'
import { convertKey, isMatchType } from './helper'
import { BreakpointMatchTypeEnum } from './interface'

declare module '@unocss/preset-mini' {
  interface Theme {
    betterBreakpoints?: BetterBreakpointsOptions
  }
}

// Reusable constants for media query construction
const PX_SUFFIX = 'px'
const MEDIA_QUERY_JOIN = ' and '
const MEDIA_TYPE_JOIN = ', '

/**
 * Transforms breakpoint configuration into CSS media query matcher
 * @param option Breakpoint configuration (object or primitive value)
 * @returns Media query string for CSS rules
 */
export function breakpointToMatcher(option: Breakpoint): string {
  // Handle primitive values directly for fast path
  if (typeof option !== 'object') {
    const size = typeof option === 'string' ? option : `${option}${PX_SUFFIX}`
    return `(min-width: ${size})`
  }

  const entries = Object.entries(option as BreakpointObj)
  const parts = Array.from({ length: entries.length })

  for (let i = 0; i < entries.length; i++) {
    const [key, size] = entries[i]
    const value = typeof size === 'string' ? size : `${size}${PX_SUFFIX}`
    parts[i] = `(${convertKey(key)}: ${value})`
  }

  return parts.join(MEDIA_QUERY_JOIN)
}

/**
 * Creates UnoCSS variant handler for responsive breakpoints
 * @returns Configured variant object for UnoCSS
 */
export function variantBreakpoints(): VariantObject<Theme> {
  // Reusable handler factory for media query variants
  const createHandler = (query: string, matcher: string): VariantHandler => ({
    matcher,
    handle: (input, next) => next({
      ...input,
      parent: `@media ${query}`,
      parentOrder: 3000,
    }),
  })

  // WeakMap preserves memory when configurations are garbage collected
  const regexCache = new WeakMap<BetterBreakpointsOptions, RegExp>()

  return {
    name: 'betterBreakpoints',
    order: -1,
    match(matcher, context) {
      const { betterBreakpoints } = context.theme
      if (!betterBreakpoints)
        return

      // Set provides O(1) lookups for breakpoint validation
      const breakpointSet = new Set(Object.keys(betterBreakpoints))
      if (!breakpointSet.size)
        return

      // Cache regex patterns per configuration to avoid recompilation
      let regex = regexCache.get(betterBreakpoints)
      if (!regex) {
        const pattern = Array.from(breakpointSet).join('|')
        regex = new RegExp(`^(?<breakpoint>${pattern}):`)
        regexCache.set(betterBreakpoints, regex)
      }

      const match = matcher.match(regex)
      const breakpoint = match?.groups?.breakpoint
      if (!breakpoint || !breakpointSet.has(breakpoint))
        return

      const option = betterBreakpoints[breakpoint]
      if (!option)
        return

      // Fast path for simple breakpoint configurations
      if (!isMatchType(option)) {
        return createHandler(breakpointToMatcher(option as Breakpoint), matcher)
      }

      // Prioritize 'all' media type handling
      const allBreakpoint = option[BreakpointMatchTypeEnum.ALL]
      if (allBreakpoint !== undefined) {
        const query = breakpointToMatcher(allBreakpoint)
        return query && createHandler(query, matcher)
      }

      // Handle combined media types with explicit control flow
      const mediaQueries = []
      const printBreakpoint = option[BreakpointMatchTypeEnum.PRINT]
      const screenBreakpoint = option[BreakpointMatchTypeEnum.SCREEN]

      // Avoid array methods for critical path performance
      if (printBreakpoint) {
        mediaQueries.push(`print and ${breakpointToMatcher(printBreakpoint)}`)
      }
      if (screenBreakpoint) {
        mediaQueries.push(`screen and ${breakpointToMatcher(screenBreakpoint)}`)
      }

      return mediaQueries.length
        ? createHandler(mediaQueries.join(MEDIA_TYPE_JOIN), matcher)
        : undefined
    },
  }
}
