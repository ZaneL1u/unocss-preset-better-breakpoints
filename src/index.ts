import type { Preset } from '@unocss/core'
import type { Theme } from '@unocss/preset-mini'
import { definePreset } from '@unocss/core'
import { variantBreakpoints } from './variant'

export const presetBetterBreakpoints = definePreset(() => {
  return {
    name: 'unocss-preset-better-breakpoints',

    variants: [variantBreakpoints()],
  } satisfies Preset<Theme>
})
