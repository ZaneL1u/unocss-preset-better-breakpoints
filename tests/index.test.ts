import { createGenerator } from '@unocss/core'
import { presetMini } from 'unocss'
// breakpoints.test.ts
import { describe, expect, it } from 'vitest'
import { presetBetterBreakpoints } from '../src'
import { breakpointToMatcher } from '../src/variant'

it('it works', async () => {
  const uno = createGenerator({
    theme: {
      betterBreakpoints: {
        md: { max: '768px' },
        sm: { max: '640px' },
      },
    },
    presets: [
      presetMini(),
      presetBetterBreakpoints(),
    ],
  })

  const { css } = await (await uno).generate('h-5 sm:h-10 md:h-20')

  expect(css).toMatchInlineSnapshot(`
    "/* layer: preflights */
    *,::before,::after{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-ring-offset-shadow:0 0 rgb(0 0 0 / 0);--un-ring-shadow:0 0 rgb(0 0 0 / 0);--un-shadow-inset: ;--un-shadow:0 0 rgb(0 0 0 / 0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgb(147 197 253 / 0.5);}::backdrop{--un-rotate:0;--un-rotate-x:0;--un-rotate-y:0;--un-rotate-z:0;--un-scale-x:1;--un-scale-y:1;--un-scale-z:1;--un-skew-x:0;--un-skew-y:0;--un-translate-x:0;--un-translate-y:0;--un-translate-z:0;--un-ring-offset-shadow:0 0 rgb(0 0 0 / 0);--un-ring-shadow:0 0 rgb(0 0 0 / 0);--un-shadow-inset: ;--un-shadow:0 0 rgb(0 0 0 / 0);--un-ring-inset: ;--un-ring-offset-width:0px;--un-ring-offset-color:#fff;--un-ring-width:0px;--un-ring-color:rgb(147 197 253 / 0.5);}
    /* layer: default */
    .h-5{height:1.25rem;}
    @media (max-width: 640px){
    .sm\\:h-10{height:2.5rem;}
    }
    @media (max-width: 768px){
    .md\\:h-20{height:5rem;}
    }"
  `)
})

describe('breakpointToMatcher', () => {
  it('handles numeric values', () => {
    expect(breakpointToMatcher(768)).toBe('(min-width: 768px)')
    expect(breakpointToMatcher(1024.5)).toBe('(min-width: 1024.5px)')
  })

  it('handles string values', () => {
    expect(breakpointToMatcher('768px')).toBe('(min-width: 768px)')
    expect(breakpointToMatcher('60em')).toBe('(min-width: 60em)')
  })

  it('handles object values', () => {
    expect(breakpointToMatcher({ min: 768, max: 1024 }))
      .toBe('(min-width: 768px) and (max-width: 1024px)')

    expect(breakpointToMatcher({ 'min-width': '768px', 'orientation': 'portrait' }))
      .toBe('(min-width: 768px) and (orientation: portrait)')
  })

  it('handles complex units', () => {
    expect(breakpointToMatcher({ min: '100vh', max: '50rem' }))
      .toBe('(min-width: 100vh) and (max-width: 50rem)')
  })
})

describe('variantBreakpoints', () => {
  it('returns undefined when no breakpoints configured', async () => {
    const uno = createGenerator()

    const { css } = await (await uno).generate('h-5 sm:h-10 md:h-20')

    expect(css).toMatchSnapshot()
  })
})
