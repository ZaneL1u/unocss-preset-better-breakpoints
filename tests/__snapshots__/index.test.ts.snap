// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`variantBreakpoints > returns undefined when no breakpoints configured 1`] = `""`;
